R/env.R

```diff
@@ -0,0 +1,702 @@
+#' @seealso \link{env_scoped}, \code{\link{env_has}()},
+#'   \code{\link{env_assign}()}.
+#' @export
+#' @examples
+#' # Get the environment of frame objects. If no argument is supplied,
+#' # the current frame is used:
+#' fn <- function() {
+#'   list(
+#'     env(call_frame()),
+#'     env()
+#'   )
+#' }
+#' fn()
+#'
+#' # Environment of closure functions:
+#' env(fn)
+#'
+#' # There is also an assignment operator:
+#' env(fn) <- env_base()
+#' env(fn)
+#'
+#'
+#' # env_new() creates by default an environment whose parent is the
+#' # current environment. Here we return a new environment that has
+#' # the evaluation environment (or frame environment) of a function
+#' # as parent:
+#' fn <- function() {
+#'   my_object <- "A"
+#'   env_new()
+#' }
+#' frame_env <- fn()
+#'
+#' # The new environment is empty:
+#' env_has(frame_env, "my_object")
+#'
+#' # But sees objects defined inside fn() by inheriting from its
+#' # parent:
+#' env_sees(frame_env, "my_object")
+#'
+#'
+#' # Create a new environment with a particular scope by setting a
+#' # parent. When inheriting from the empty environment, the
+#' # environment will have no object in scope at all:
+#' env <- env_new(env_empty())
+#' env_sees(env, "lapply")
+#'
+#' # The base package environment is often a good default choice for a
+#' # parent environment because it contains all standard base
+#' # functions:
+#' env <- env_new(env_base())
+#' env_sees(env, "lapply")
+#'
+#' # Note that all other package environments inherit from env_base()
+#' # as well:
+#' env <- env_new(env_package("rlang"))
+#' env_sees(env, "env_sees")
+#' env_sees(env, "lapply")
+#'
+#'
+#' # Get the parent environment with env_next():
+#' env_next(env_global())
+#'
+#' # Or the tail environment with env_tail():
+#' env_tail(env_global())
+#'
+#'
+#' # By default, env_next() returns the parent environment of the
+#' # current evaluation frame. If called at top-level (the global
+#' # frame), the following two expressions are equivalent:
+#' env_next()
+#' env_next(env_global())
+#'
+#' # This default is more handy when called within a function. In this
+#' # case, the enclosure environment of the function is returned
+#' # (since it is the parent of the evaluation frame):
+#' enclos_env <- env_new()
+#' fn <- with_env(enclos_env, function() env_next())
+#' identical(enclos_env, fn())
+env <- function(env = env_caller()) {
+  UseMethod("env")
+}
+
+#' @rdname env
+#' @export
+env.function <- function(env = env_caller()) {
+  environment(env)
+}
+#' @rdname env
+#' @export
+env.formula <- function(env = env_caller()) {
+  attr(env, ".Environment")
+}
+#' @rdname env
+#' @export
+env.frame <- function(env = env_caller()) {
+  env$env
+}
+#' @rdname env
+#' @export
+env.environment <- function(env = env_caller()) {
+  env
+}
+#' @rdname env
+#' @export
+env.default <- function(env = env_caller()) {
+  # Default argument env_caller() gets dispatched here:
+  if (is_env(env)) {
```

Comment:
Missing arguments get dispatched to `.default`, so we check for the default environment.